TypeScript rules

molant opened the pull request:
<!--

Read our pull request guide:
https://sonarwhal.com/docs/contributor-guide/contributing/pull-requests/

For the following items put an "x" between the square brackets
(i.e. [x]) if you completed the associated item.

-->

## Pull request checklist

Make sure you:

- [X] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [X] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [X] Added/Updated related documentation.
- [X] Added/Updated related tests.

## Short description of the change(s)

This PR adds a couple new rules for TypeScript. Couple things that I'd like to do first this PR is merged:

* Find a better name for the `target` rule
* Complete the `target` support matrix, `browserslist` returns a lot of browsers in some of the queries and this should be more complete
* Maybe add what browser(s) are affecting the `target`? 

In any case, this is ready to start looking in case there's any major change required.

molant commented:
This PR also fixes some stuff that could be merged before (first 3 commits)

alrra commented:
> This PR also fixes some stuff that could be merged before (first 3 commits)

Done.

sarvaje commented on packages/rule-typescript-config/src/target.ts:
```diff
@@ -0,0 +1,235 @@
+            }
+
+            // Can't store `typeof` or otherwise TypeScript complains about `includes` not being on `number`
+            if (typeof version1 === 'string' && typeof version2 === 'string') {
```
IMO, this is to complex (or I'm missing something). Why don't you just parse the versions?
```
if(typeof v1 === 'string') {
v1 = getMajor(v1)
}
if(typeof v2 === 'string') {
v2 = getMajor(v2)
}
return isOlder(v1, v2)
```

molant replied:
Browser versions are complicated...
We could be comparing:
* `safari 10.1` and `safari 10.0`
* `android 3.0.1` and `android 4.4.3`
* In some cases we get a range as a version: `android 4.4.3-4.4.4`.

I should put some comments or try to refactor the code to make it clearer.

sarvaje replied:
but at the end, you just compare the major version, isn't it?

molant replied:
I should compare the mayor only if the other tests have failed.

sarvaje replied:
ah, ok

sarvaje commented on packages/rule-typescript-config/src/target.ts:
```diff
@@ -0,0 +1,235 @@
+             * If there's one in the compat matrix newer than the ones
+             * provided, it will fail.
+             */
+            versions.forEach((version) => {
```
shouldn't  this be a `reduce`?

molant replied:
Yes, I can probably change it to `reduce`

sarvaje commented on packages/rule-typescript-config/src/target.ts:
```diff
@@ -0,0 +1,235 @@
+         * * ES6 --> ES2015
+         */
+        const getConfiguredTarget = (config: TypeScriptConfig): string => {
+            const target = config && config.compilerOptions && config.compilerOptions.target || 'es3';
```
I think you don't need to check if `config` exists, if you receive an `parse::typescript-config` then the config has to exist.

sarvaje commented on packages/rule-typescript-config/src/target.ts:
```diff
@@ -0,0 +1,235 @@
+        const getConfiguredTarget = (config: TypeScriptConfig): string => {
+            const target = config && config.compilerOptions && config.compilerOptions.target || 'es3';
+
+            return Targets.get(target.toLowerCase());
```
`Target` is not necessary, you can return `target.toLowerCase()` directly

molant replied:
target could be `es6` and I need to transform it to `es2015` (thus my issue about normalizing things)

sarvaje replied:
You are right, I missed that value.

molant commented on packages/rule-typescript-config/README.md:
```diff
@@ -50,9 +62,171 @@ The configuration is valid:
+
+The matrix is as follows:
+
+| Target | Chrome | Edge | Firefox | IE | Safari |
```
Nope, see the comment in `Short description of the change(s)`

sarvaje replied:
ups, I missed that

molant commented on packages/rule-typescript-config/docs/target.md:
```diff
@@ -0,0 +1,113 @@
+| Target | Chrome | Edge | Firefox | IE | Safari |
+| ------ | ------ | ---- | ------- | -- | ------ |
+|  ES3   |   ✔    |  ✔  |    ✔    | ✔ |    ✔   |
+|  ES5   |   57   |  12  |   45    | 9 |    8    |
```
I'm sure the values for chrome are not the right ones. Need to recheck the kangax tables and update this accordingly.

molant commented:
@sarvaje @alrra, I've updated the browser support matrix and added a few more links. Let me know if you have any other feedback or otherwise I'll merge it as soon as I can.
